### ISSUE TITLE
img-weights

### ISSUE THREAD

**Issue by yangjinghit** [ultralytics/yolov5#1238]
## ❔Question
parser.add_argument('--img-weights', action='store_true', help='use weighted image selection for training')
in order to make --iimg-weights work,  what else I need to do? 
dataset = LoadImagesAndLabels(path, imgsz, batch_size,
                                    augment=augment,  # augment images
                                    hyp=hyp,  # augmentation hyperparameters
                                    rect=rect,  # rectangular training
                                    cache_images=cache,
                                    single_cls=opt.single_cls,
                                    stride=int(stride),
                                    pad=pad),
 should I add an extra param image_weights=True??

**Comment by yangjinghit** [comment]
@glenn-jocher

**Comment by yangjinghit** [comment]
w = model.class_weights.cpu().numpy() * (1 - maps) ** 2  # class weights
                image_weights = labels_to_image_weights(dataset.labels, nc=nc, class_weights=w)
                dataset.indices = random.choices(range(dataset.n), weights=image_weights,
                                                 k=dataset.n)  # rand weighted idx

I did not find anywhere show how to use dataset.indices in above codes

**Comment by glenn-jocher** [comment]
@yangjinghit to use image weighting you simply pass the argument during training:
`python train.py --img-weights`

**Comment by MaxwellHogan** [comment]
@glenn-jocher Hey Bud, could you summarise what img-weights does?

is it similar to threshold-moving:
https://machinelearningmastery.com/threshold-moving-for-imbalanced-classification/

**Comment by glenn-jocher** [comment]
@MaxwellHogan `--image-weights` samples images from the training set weighted by their inverse mAP from the previous epoch's testing (rather than sampling the images uniformly as in normal training). This will result in images with a high content of low-mAP objects being selected with higher likelihood during training.

This is not the same as the method in your link.

**Comment by MaxwellHogan** [comment]
@glenn-jocher thanks for the prompt feedback!

**Comment by Barry04** [comment]
能不能描述一下参数--image-weights的作用，会对训练的结果有影响吗？是否建议开启这个参数？

**Comment by thusinh1969** [comment]
> 



> @MaxwellHogan `--image-weights` samples images from the training set weighted by their inverse mAP from the previous epoch's testing (rather than sampling the images uniformly as in normal training). This will result in images with a high content of low-mAP objects being selected with higher likelihood during training.
> 
> This is not the same as the method in your link.

Oh, would this cause overfit to minor classes I am a bit afraid !??! Thanks.

**Comment by XueZ-phd** [comment]
> @MaxwellHogan `--image-weights` samples images from the training set weighted by their inverse mAP from the previous epoch's testing (rather than sampling the images uniformly as in normal training). This will result in images with a high content of low-mAP objects being selected with higher likelihood during training.
> 
> This is not the same as the method in your link.

Great work! Can I regard the 'image-weights' as an approach to long-tail distributed data set?

**Comment by glenn-jocher** [comment]
@XueZ-phd thank you for your feedback and question! Yes, you are correct that `--img-weights` is one approach to handling long-tail distributed datasets. By sampling training images based on the inverse mAP from the previous epoch's testing, `--img-weights` can prioritize training of low-mAP object classes, leading to better detection performance on these underrepresented classes.

**Comment by XueZ-phd** [comment]
> @XueZ-phd thank you for your feedback and question! Yes, you are correct that `--img-weights` is one approach to handling long-tail distributed datasets. By sampling training images based on the inverse mAP from the previous epoch's testing, `--img-weights` can prioritize training of low-mAP object classes, leading to better detection performance on these underrepresented classes.

Thanks! It is really a convenient and elegant operation!

**Comment by glenn-jocher** [comment]
@XueZ-phd You're welcome! I'm glad that you find the `--img-weights` approach convenient and elegant for handling long-tail distributed datasets. It can indeed be an effective way to prioritize training of underrepresented classes and improve overall object detection performance. Let us know if you have any additional questions or feedback.

**Comment by XueZ-phd** [comment]
@glenn-jocher It is really an effective way to improve the detection performance of underrepresented classes by oversampling these classes, but it may be less effective for all classes. I have experimentally found that enabling `--img-weights` boosts the accuracy of these underrepresented classes. However, an unwanted phenomenon is that the performance of classes with plentiful instances decreases. I guess this phenomenon results from the model overfitting the classes with fewer instances while not being able to generalize to detect classes with more instances (I am not sure about this guess).

For example, in my experiments, there are 6 classes with the following numbers of instances: "people" (7966), "car" (12880), "bus" (513), "motorcycle" (361), "lamp" (1785), and "truck" (719). The performance comparison between enabling `--img-weights` and not enabling it is shown below (unfortunately, I cannot upload images to show the results, otherwise it would be more readable).
```
Not Enabling --img-weights
- people: 0.829
- car: 0.926
- bus: 0.908
- motorcycle: 0.781
- lamp: 0.825
- truck: 0.856
- all classes: 0.854 mAP@0.5
```

```
Enabling --img-weights
- people: 0.817 (-1.2%)
- car: 0.919 (-0.7%)
- bus: 0.895 (-1.3%)
- motorcycle: 0.822 (+4.1%)
- lamp: 0.860 (+3.5%)
- truck: 0.835 (-2.1%)
- all classes: 0.858 mAP@0.5 (+0.4%)
```
* Both experiments are conducted with the same random seed, torch==1.12.1, and the experiment is reproducible.

**Comment by glenn-jocher** [comment]
Thank you for bringing up this issue @XueZ-phd! You are correct that enabling --img-weights can lead to better detection performance on underrepresented classes by oversampling them. However, this approach may not be equally effective on all classes and can even hurt the performance of over-represented classes, as you have observed in your experiments. Your guess that this phenomenon results from the model overfitting to the underrepresented classes while not being able to generalize to the over-represented ones seems reasonable. This is a common issue in object detection and there are several strategies that can be explored to mitigate it. Some possible solutions could be to adjust the loss weights of different classes, use a balanced sampling approach, or add data augmentation techniques that increase the variability of the over-represented classes. These are just some suggestions and there could be other approaches to handle this issue as well.

**Comment by XueZ-phd** [comment]
@glenn-jocher Thank you very much for agreeing with my guess and for your valuable suggestion to handle the long-tail issue. I will try these strategies in my experiments.

**Comment by glenn-jocher** [comment]
You are welcome @XueZ-phd! I'm glad that you found my suggestions helpful and will try implementing them in your experiments. Please feel free to reach out if you have any further questions or feedback. Good luck with your experiments!

**Comment by pathikg** [comment]
So @glenn-jocher , 
Since sampling takes place inside the dataloader, where for each sampled batch we're giving higher priority to the classes having low mAP thereby increasing their distribution in the training set
Why don't we have the same for yolov8 ?

**Comment by glenn-jocher** [comment]
@pathikg, 

The `--img-weights` functionality in YOLOv5 is designed to sample images from the training set using inverse mAP weights, giving higher priority to classes with lower mAP. This helps to balance the distribution of the underrepresented classes during training.

Regarding your question about YOLOv8, it doesn't have a built-in option like `--img-weights` for weighted image selection. However, you can still implement a similar approach manually by modifying the dataloader to assign weights to images based on class mAP or other criteria.

Let me know if you have any further questions!

Glenn Jocher

**Comment by jeffrey-cochran** [comment]
Okay, just want to make sure I'm understanding you @glenn-jocher. If I wanted to implement, for example, a multiclass adaboost ensemble with yolov8, there's no existing option for me to pass image weights--instead, I'll have to modify the dataloader to sample the images with the frequency corresponding to their weight?

**Comment by glenn-jocher** [comment]
@jeffrey-cochran yes, that's correct. In YOLOv8, there is no built-in option like `--img-weights` in YOLOv5 to handle weighted image selection. If you want to implement a multiclass adaboost ensemble with YOLOv8 and use image weights, you would need to modify the dataloader to sample images based on their weight (frequency). This would involve assigning weights to each image in the dataset and adjusting the sampling logic accordingly to ensure images are selected with the appropriate frequency based on their weight.

Let me know if you have any further questions or need assistance with the implementation.

Glenn Jocher

**Comment by geolvr** [comment]
> @MaxwellHogan `--image-weights` samples images from the training set weighted by their inverse mAP from the previous epoch's testing (rather than sampling the images uniformly as in normal training). This will result in images with a high content of low-mAP objects being selected with higher likelihood during training.
> 
> This is not the same as the method in your link.

@glenn-jocher 
I have some questions about whether the weights are actually based on the testing mAP in the previous epoch. In the latest version （7.0), I can't find the corresponding codes.

The codes associated with _image_weights_ is in _train.py_:

`if opt.image_weights:
    cw = model.class_weights.cpu().numpy() * (1 - maps) ** 2 / nc  # class weights
    iw = labels_to_image_weights(dataset.labels, nc=nc, class_weights=cw)  # image weights
    dataset.indices = random.choices(range(dataset.n), weights=iw, k=dataset.n)  # rand weighted idx`

It seems that the training set is simply sampled with a fixed weight (which is derived from the class distribution of the labels in the training set) before each epoch training starts, rather than the test mAP based on the previous epoch.

Is my understanding wrong?

**Comment by glenn-jocher** [comment]
@geolvr Thank you for bringing this to my attention. You are correct. The weights used with `--image-weights` are derived from class distribution and not from the test mAP of the previous epoch. I misspoke earlier and apologize for the confusion. The weights are indeed derived from the class distribution of the labels in the training set, as shown in the code snippet you provided from `train.py`.

Thank you for pointing this out, and let me know if you have any further questions or concerns.

**Comment by fungtion** [comment]
@glenn-jocher you are right, `--image-weights` are indeed derived from the ap of every single class. @geolvr map updates after every validation [here](https://github.com/ultralytics/yolov5/blob/d07d0cf633b8b8bca90509d4c6183e62d3ebaca6/train.py#L426)

**Comment by glenn-jocher** [comment]
Thank you for pointing that out! You're correct that the `--image-weights` option dynamically updates based on class performance, making use of the AP from validation to adjust the sampling. This ensures that underperforming classes are given more emphasis during training, aiding in a more balanced learning process.

I appreciate your clarification on this matter! If you have any more questions or insights, feel free to share. Happy coding! 😊

**Comment by Yosu26** [comment]
is this technique based on OHEM (Online Hard Example Mining) concept? or is there a more broad term or technique for img-weights? 
Thank you in advance!

**Comment by glenn-jocher** [comment]
Hello @Yosu26,

Thank you for your question! The `--image-weights` technique in YOLOv5 is somewhat related to the concept of Online Hard Example Mining (OHEM), but it is not exactly the same. 

OHEM focuses on selecting the hardest examples (i.e., those that the model finds most difficult to classify correctly) during training to improve the model's performance. In contrast, the `--image-weights` option in YOLOv5 dynamically adjusts the sampling of images based on the performance of each class. Specifically, it gives higher priority to classes with lower average precision (AP) from the previous validation epoch, ensuring that underperforming classes are more frequently sampled and thus receive more attention during training.

This approach helps to balance the training process by focusing on classes that need more improvement, which can be particularly useful in datasets with imbalanced class distributions.

If you have any further questions or need additional clarification, feel free to ask. We're here to help! 😊
